// @urql/exchange-graphcache/dist/urql-exchange-graphcache-default-storage.mjs
import { _ as e } from "./5301ccd2.mjs";

function getRequestPromise(e) {
  return new Promise((function(t, r) {
    e.onerror = function() {
      r(e.error);
    };
    e.onsuccess = function() {
      t(e.result);
    };
  }));
}

function getTransactionPromise(e) {
  return new Promise((function(t, r) {
    e.onerror = function() {
      r(e.error);
    };
    e.oncomplete = t;
  }));
}

function makeDefaultStorage(t) {
  if (!t) {
    t = {};
  }
  var r = t.idbName || "graphcache-v4";
  var n = "entries";
  var a = "metadata";
  var i = Object.create(null);
  var o = Math.floor((new Date).valueOf() / 864e5);
  var u = o - (t.maxAge || 7);
  var c = indexedDB.open(r, 1);
  var s = getRequestPromise(c);
  c.onupgradeneeded = function() {
    c.result.createObjectStore(n);
    c.result.createObjectStore(a);
  };
  function serializeEntry(e) {
    return e.replace(/:/g, "%3a");
  }
  function deserializeEntry(e) {
    return e.replace(/%3a/g, ":");
  }
  function deserializeBatch(e) {
    var t = {};
    var r = "", n = "", a = "", i = 0, o = 0;
    while (o < e.length) {
      a = "";
      while (":" !== (r = e[o++]) && r) {
        a += r;
      }
      if (i) {
        t[n] = deserializeEntry(a) || void 0;
        i = 0;
      } else {
        n = deserializeEntry(a);
        i = 1;
      }
    }
    return t;
  }
  return {
    clear: function clear() {
      return s.then((function(e) {
        var t = e.transaction([ a, n ], "readwrite");
        t.objectStore(a).clear();
        t.objectStore(n).clear();
        return getTransactionPromise(t);
      }));
    },
    readMetadata: function readMetadata() {
      return s.then((function(e) {
        return getRequestPromise(e.transaction(a, "readonly").objectStore(a).get(a));
      }), (function() {
        return null;
      }));
    },
    writeMetadata: function writeMetadata(e) {
      s.then((function(t) {
        return getRequestPromise(t.transaction(a, "readwrite").objectStore(a).put(e, a));
      }), (function() {}));
    },
    writeData: function writeData(t) {
      e(i, t);
      var toUndefined = function() {
        return;
      };
      return s.then((function(e) {
        return getRequestPromise(e.transaction(n, "readwrite").objectStore(n).put(function serializeBatch() {
          var e = "";
          for (var t in i) {
            var r = i[t];
            e += serializeEntry(t);
            e += ":";
            if (r) {
              e += serializeEntry(r);
            }
            e += ":";
          }
          return e;
        }(), o));
      })).then(toUndefined, toUndefined);
    },
    readData: function readData() {
      var t = [];
      return s.then((function(r) {
        var a = r.transaction(n, "readwrite");
        var c = a.objectStore(n);
        (c.openKeyCursor || c.openCursor).call(c).onsuccess = function() {
          if (this.result) {
            var r = this.result.key;
            if ("number" != typeof r || r < u) {
              c.delete(r);
            } else {
              var n = c.get(r);
              var a = t.length;
              t.push("");
              n.onsuccess = function() {
                var u = "" + n.result;
                if (r === o) {
                  e(i, deserializeBatch(u));
                }
                t[a] = u;
              };
            }
            this.result.continue();
          }
        };
        return getTransactionPromise(a);
      })).then((function() {
        return deserializeBatch(t.join(""));
      }), (function() {
        return i;
      }));
    },
    onOnline: function onOnline(e) {
      window.addEventListener("online", (function() {
        e();
      }));
    }
  };
}

export { makeDefaultStorage };
